rusb-async/src/transfer/isochronous.rs

```diff
@@ -0,0 +1,236 @@
+    C: AsyncUsbContext,
+{
+    fn fill(&mut self, waker: Waker) -> Result<()> {
+        let length = if self.endpoint & LIBUSB_ENDPOINT_DIR_MASK == LIBUSB_ENDPOINT_OUT {
+            // for OUT endpoints: the currently valid data in the buffer
+            self.buffer.len()
+        } else {
+            // for IN endpoints: the full capacity
+            self.buffer.capacity()
+        };
+
+        let length: i32 = length
+            .try_into()
+            .map_err(|_| Error::Other("Invalid buffer length"))?;
+
+        let packet_lengths = (length / self.kind.iso_packets)
+            .try_into()
+            .map_err(|_| Error::Other("Invalid iso packets length"))?;
+
+        let user_data = Box::into_raw(Box::new(TransferUserData::new(waker))).cast();
+
+        unsafe {
+            ffi::libusb_fill_iso_transfer(
+                self.ptr.as_ptr(),
+                self.dev_handle.as_raw(),
+                self.endpoint,
+                self.buffer.as_mut_ptr(),
+                length,
+                self.kind.iso_packets,
+                Self::transfer_cb,
+                user_data,
+                0,
+            );
+
+            ffi::libusb_set_iso_packet_lengths(self.ptr.as_ptr(), packet_lengths);
+        }
+
+        Ok(())
+    }
+}
+
+/// Separate implementation for [`IsochronousTransfer`] since
+/// the buffer represents multiple packets.
+///
+/// As opposed to other transfers, this returns an iterator of
+/// [`Result`], one per isochronous packet, each containing a
+/// slice of the buffer representing the packet data.
+impl<C> CompleteTransfer for IsochronousTransfer<C>
+where
+    C: AsyncUsbContext,
+{
+    type Output = IsochronousBuffer;
+
+    fn consume_buffer(&mut self, mut buffer: Vec<u8>) -> Result<Self::Output> {
+        let len = self.transfer().length.try_into().unwrap();
+        unsafe { buffer.set_len(len) };
+
+        let packet_descriptors = self
+            .packet_descriptors()
+            .iter()
+            .map(TryFrom::try_from)
+            .collect::<Result<_>>()?;
+
+        Ok(IsochronousBuffer {
+            packet_descriptors,
+            buffer,
+        })
+    }
+}
+
+/// Isochronous packet descriptor.
+#[derive(Debug, Copy, Clone)]
+struct IsochronousPacketDescriptor {
+    length: usize,
+    actual_length: usize,
+    status: libc::c_int,
+}
+
+/// We convert from a reference because that's what we can safely get
+/// from the underlying transfer pointer.
+impl TryFrom<&libusb_iso_packet_descriptor> for IsochronousPacketDescriptor {
+    type Error = Error;
+
+    fn try_from(value: &libusb_iso_packet_descriptor) -> std::result::Result<Self, Self::Error> {
+        let length = value
+            .length
+            .try_into()
+            .map_err(|_| Error::Other("Invalid isochronous packet length"))?;
+        let actual_length = value
+            .length
+            .try_into()
+            .map_err(|_| Error::Other("Invalid isochronous packet actual length"))?;
```

Comment:
fix: you must use `actual_length` instead of `length` here, otherwise all iso packets that don't use the full capacity will have trailling junk data.

```suggestion
        let actual_length = value
            .actual_length
            .try_into()
            .map_err(|_| Error::Other("Invalid isochronous packet actual length"))?;
```